src/main/java/org/broadinstitute/hellbender/tools/exome/ExomeReadCounts.java

```diff
@@ -0,0 +1,881 @@
+            }
+        }
+
+        /**
+         * Returns the name for a column given its index.
+         *
+         * <p>
+         *     Indexes are 0-based.
+         * </p>
+         * @param index the column index.
+         * @return never {@code null}.
+         */
+        protected abstract String columnName(final int index);
+
+        /**
+         * Composes the list of count column names.
+         *
+         * <p>
+         *     The result list would have exactly <code>{@link #columnCount()}</code> elements.
+         * </p>
+         *
+         * @return never {@code null}.
+         */
+        private List<String> composeCountColumns() {
+            final int columnCount = columnCount();
+            final Set<String> resultSet = new LinkedHashSet<>(columnCount);
+            final String[] result = new String[columnCount];
+            for (int i = 0; i < columnCount; i++) {
+                resultSet.add(result[i] = columnName(i));
+            }
+            if (resultSet.size() != columnCount) {
+                throw new UserException.BadArgumentValue(COUNT_COLUMN_FULL_NAME,
+                        "count column name pattern results in more than one column with the same name: " +
+                                String.join(", ", result));
+            }
+            return Collections.unmodifiableList(Arrays.asList(result));
+        }
+
+        /**
+         * Returns the count column index corresponding to a read.
+         * @param read the query read.
+         * @return a value from 0 to <code>{@link #columnCount()} - 1</code>, or -1 if the value is to be discarded.
+         */
+        protected abstract int columnIndex(SAMRecord read);
+
+        /**
+         * Returns the count column headers.
+         * @return never {@code null}, an immutable list of header where
+         */
+        private List<String> columnNames() {
+            return columnNames;
+        }
+
+        /**
+         * Returns the number of count columns headers.
+         * @return never {@code null}.
+         */
+        protected abstract int columnCount();
+    }
+
+    /**
+     * Counts class to be use when {@link #groupBy} is {@code GroupBy#COHORT}.
+     */
+    protected static final class CohortCountColumns extends CountColumns {
+
+
+        /**
+         * Creates a new count-counts instance given the reference to the embedding tool instance.
+         * @throws IllegalArgumentException if {@code tool} is {@code null}.
+         * @throws IllegalStateException if the current state of {@code tool} is not compatible with
+         *    this count-column manager type. See {@link #checkToolState(ExomeReadCounts)}.
+         */
+        private CohortCountColumns(final ExomeReadCounts tool) {
+            super(tool);
+        }
+
+        @Override
+        protected void checkToolState(final ExomeReadCounts tool) {
+            super.checkToolState(tool);
+            if (tool.groupBy != GroupBy.COHORT) {
```

Comment:
Is a paranoia check I will simply removed; it only can be trigger by a bug in the private/compilation-time resoved section of the class.